# 項目：評估並清理英國電商公司銷售數據

## 分析目標

此分析旨在根據市場銷售數據，挖掘暢銷產品，從而製定更好的市場策略來提升績效。

本實用新型的目的在於練習資料乾淨整潔，並且根據練習結果對資料進行清洗，從而得到可供分析的資料。

## 簡介

原始資料集記錄了一家英國線上零售公司在2010年12月1日至2011年12月9日期間的所有交易情況，涵蓋了該公司在全球不同國家和地區的業務資料。該公司主要銷售涵蓋各個場景的禮品，包括但不限於生日禮品、婚禮紀念品、聖誕禮品等等。該公司的客戶群主要包括批發商和個人消費者，其中批發商佔了相當大的比例。

資料每列的含義如下：
- `InvoiceNo`: 發票號碼。 6位數，作為交易的唯一識別碼。如果這個代碼以字母「c」開頭，表示這筆交易被取消。
- `StockCode`: 產品代碼。 5位數，作為產品的唯一識別碼。
- `Description`: 產品名稱。
- `Quantity`: 產品在交易中的數量。
- `InvoiceDate`: 發票日期和時間。交易發生的日期和時間。
- `UnitPrice`: 單價。價格單位為英鎊（£）。
- `CustomerID`: 客戶編號。 5位數，作為客戶的唯一識別碼。
- `Country`: 國家名稱。客戶所居住的國家的名稱。

## 讀取數據

導入資料分析所需的函式庫，並透過Pandas的`read_csv`函數，將原始資料檔"e_commerce.csv"裡的資料內容，解析為DataFrame，並賦值給變數`original_data`。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 評估數據

在這一部分，我將對上一部分建立的`original_data`這個DataFrame所包含的資料進行評估。

評估主要從兩個方面進行：結構和內容，即整齊度和乾淨度。資料的結構性問題指不符合「每列是一個變量，每行是一個觀察值，每個單元格是一個值」這三個標準，資料的內容性問題包括存在遺失資料、重複資料、無效資料等。

### 評估數據整齊度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
485660    577615     21682     LARGE MEDINA STAMPED METAL BOWL          4   
423518    573150     84375        SET OF 20 KIDS COOKIE CUTTERS         3   
206538    554942    71406C                BLACK ORANGE SQUEEZER         9   
195868    553752     21106  CREAM SLICE FLANNEL CHOCOLATE SPOT          6   
269751    560504     22553               PLASTERS IN TIN SKULLS        12   
45334     540255    84251C          GREETING CARD, TWO SISTERS.         4   
421911    573029     22277    COSMETIC BAG VINTAGE ROSE PAISLEY         3   
213885    555558     82581                    TOILET METAL SIGN         3   
89335     543900     17003                  BROCADE RING PURSE         36   
321075    565124     23243  SET OF TEA COFFEE SUGAR TINS PANTRY         8   

             InvoiceDate  UnitPrice  CustomerID         Country  
485660  11/21/2011 10:22       4.95     12658.0         Germany  
423518  10/27/2011 20:07       2.10     15014.0  United Kingdom  
206538   5/27/2011 14:01       0.42     17841.0  United Kingdom  
195868   5/19/2011 10:20       2.95     14896.0  United Kingdom  
269751   7/19/2011 11:47       1.65     13081.0  United Kingdom  
45334     1/5/2011 16:50       0.43         NaN  United Kingdom  
421911  10/27/2011 13:33       2.10     14971.0  United Kingdom  
213885    6/5/2011 14:06       0.55     14159.0  United Kingdom  
89335    2/14/2011 12:11       0.29     12621.0         Germany  
321075    9/1/2011 12:09       4.95     12436.0          Norway

從抽樣的10行資料資料來看，資料符合“每列是一個變量，每行是一個觀察值，每個儲存格都是一個值”，具體來看每行是關於某商品的一次交易，每列是交易相關的各個變量，因此不存在結構性問題。

### 評估數據乾淨度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


從輸出結果來看，資料共有541909個觀察值，而`Description`、`CustomerID`變數存在缺失值。

此外，`InvoiceDate`的資料型別應為日期，`CustomerID`的資料型別應為字串，應進行資料格式轉換。

#### 評估缺失數據

在了解`Description`存在缺失值後，根據條件提取出缺失觀察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454筆交易資料缺失`Description`變數值。

從輸出結果來看，這些缺失`Description`的交易數據，`UnitPrice`都是0。為了驗證猜想，我們增加篩選條件，看是否存在`Description`變數缺失且`UnitPrice`不為0的資料。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

篩選出來結果數為0條，說明缺失`Description`值的數據，同時也不具備有效的`UnitPrice`值。

`Description`表示產品名稱，`UnitPrice`表示產品單價，都是進行後續商品交易分析的重要變數。如果它們同時缺失/無效，我們認為資料無法提供有效含義，因此這些後續可以被刪除。

`CustomerID`變數同樣存在缺失值，因此也根據條件擷取缺失觀察值。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`表示顧客編號，並非分析暢銷商品的必要變數。而從輸出結果來看，有些`CustomerID`缺少的交易資料仍然有效，因此保留此變數為空的觀察值。

#### 評估重複數據

根據資料變數的意義來看，雖然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一識別符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重複，不同交易可以包含同件商品，因此`StockCode`可以存在重複，顧客可以進行多次交易或下單多個商品，因此`Customer`也可以存在重複。

那麼針對此資料集，我們無需評估重複資料。

#### 評估不一致數據

不一致資料可能存在於`Country`變數中，我們要查看是否存在多個不同值指涉同一國家的情況。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


從`Country`變數值來看，`"USA"`、`"United States"`皆在表示美國，`"United Kingdom"`、`"UK"`、`"U.K."`皆在表示英國，因此應該在清洗步驟對這些值進行統一，只保留一個指代值。

#### 評估無效或錯誤數據

可以透過DataFrame的`describe`方法，對數值統計資訊進行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

從輸出結果來看，`Quantity`和`UnitPrice`存在負數，會對後續數值分析產生影響。

因此，我們先篩選出`Quantity`數值為負數的觀察值，進一步評估其意義。

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

從篩選結果來看，`Quantity`變數為負數的觀察值，`InvoiceNo`似乎都以C開頭，表示訂單被取消。

但此猜測不一定準確，為了驗證猜想，我們增加篩選條件，看是否存在`Quantity`變數為負數且`InvoiceNo`不以C開頭的觀察值。

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被證實錯誤，因為還存在`Quantity`變數為負數且`InvoiceNo`不以C開頭的觀察值。

但根據上述輸出結果，這些篩選出的觀察值的`UnitPrice`觀察值均為0，因此增加`UnitPrice`的條件進行驗證。

In [13]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根據輸出結果，猜想得到驗證，當`Quantity`變數為負數的時候，觀察值滿足以下條件之一：
- `InvoiceNo`以C開頭，表示訂單被取消
- `UnitPrice`為0，說明單價為0英鎊

這些交易數據均不是有效成交數據，不貢獻銷售，不在後續分析範圍內，因此我們將在數據清理步驟，將`Quantity`為負數的觀察值刪除。

接下來篩選出`UnitPrice`為負數的觀察值。

In [14]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

從輸出結果來看，`UnitPrice`為負數的觀察值都是壞帳調賬，不屬於實際商品交易數據，因此也在數據清理步驟中也將其刪除。

## 清理數據

根據前面評估部分所得到的結論，我們需要進行的資料清理包括：
- 把`InvoiceDate`變數的資料類型轉換為日期時間
- 把`CustomerID`變數的資料型別轉換為字串
- 把`Description`變數缺失的觀察值刪除
- 把`Country`變數值`"USA"`替換為`"United States"`
- 把`Country`變數值`"UK"`、`"U.K."`替換為`"United Kingdom"`
- 將`Quantity`變數值為負數的觀察值刪除
- 把`UnitPrice`變數值為負數的觀察值刪除

為了區分開經過清理的資料和原始的數據，我們建立新的變數cleaned_data，讓它為original_data複製出的副本。我們之後的清理步驟都會被運用在cleaned_data上。

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [16]:
cleaned_data["InvoiceDate"]

0          12/1/2010 8:26
1          12/1/2010 8:26
2          12/1/2010 8:26
3          12/1/2010 8:26
4          12/1/2010 8:26
               ...       
541904    12/9/2011 12:50
541905    12/9/2011 12:50
541906    12/9/2011 12:50
541907    12/9/2011 12:50
541908    12/9/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

把`InvoiceDate`變數的資料型別轉換為日期時間型別：

In [17]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`ustomerID`變數的資料型別轉換為字串：

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerID`變數值結尾的`".0"`刪除，因為客戶ID不需要小數點：

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`變數缺少的觀察值刪除，並查看刪除後該列空缺值個數和：

In [20]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [21]:
cleaned_data["Description"].isnull().sum()

0

把`Country`變數值`"USA"`替換為`"United States"`，並檢查替換後`"USA"`變數值個數：

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [23]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把`Country`變數值`"UK"`、`"U.K."`替換為`"United Kingdom"`，並檢查替換後`"UK"`和`"U.K."`變數值個數：

In [24]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})


In [25]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


將`Quantity`變數值為負數的觀察值刪除，並檢查替換後`Quantity`變數值為負數的個數：

In [26]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [27]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

將`UnitPrice`變數值為負數的觀察值刪除，並檢查替換後`UnitPrice`變數值為負數的個數：

In [28]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [29]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理後的數據

完成資料清理後，把乾淨整齊的資料存到新的檔案裡，檔案名稱為`e_commerce_cleaned.csv`。

In [30]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [31]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [32]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\Admin\AppData\Local\Temp\ipykernel_7208\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom